gemrb/core/DisplayMessage.cpp

```diff
@@ -190,28 +188,26 @@ void DisplayMessage::DisplayString(String text, const Color &color, Scriptable *
 	}
 }
 
-std::vector<Color> DisplayMessage::GetAllColors() const
+std::vector<std::string> DisplayMessage::GetAllColors() const
 {
-	std::vector<Color> auxiliaryColors;
-	if (auxiliaryColors.empty()) {
-		AutoTable colorTable = gamedata->LoadTable("colors", true);
-		assert(colorTable);
-		for (TableMgr::index_t r = 0; r < colorTable->GetRowCount(); r++) {
-			ieDword c = colorTable->QueryFieldUnsigned<ieDword>(r, 0);
-			auxiliaryColors.push_back(Color(c));
-		}
+	std::vector<std::string> auxiliaryColors;
+	AutoTable colorTable = gamedata->LoadTable("colors", true);
+	assert(colorTable);
+	for (TableMgr::index_t r = 0; r < colorTable->GetRowCount(); r++) {
+		auxiliaryColors.push_back(colorTable->GetRowName(r));
 	}
 	return auxiliaryColors;
 }
 
 Color DisplayMessage::GetColor(GUIColors color) const
 {
-	int colortInt = static_cast<int>(color) - 1;
-	int size = static_cast<int>(colors.size());
-	if (size < colortInt) {
-		return ColorRed;
+	
+	int colorInt = static_cast<int>(color) - 1;
+	if (static_cast<int>(GUIColorNames.size()) < colorInt) {
+		// It never should happend this, but just for been defensive.
+		colorInt = 0;
 	}
-	return colors.at(colortInt);
+	return gamedata->GetColor(GUIColorNames[colorInt]);
```

Comment:
Let's use `std::map<GUIColors, std::string>` then for GUIColorNames, so this function can be simplified or removed altogether. Because it's a map, it also means you don't have to check for non-existing keys, as new entries would get created automatically and eventually the default color from the table would be used.

Reply:
@lynxlynxlynx this will not be the same as the function **gamedata->GetColor** at the end ?

Reply:
No, the gamedata version will still be the main one to get the precached Color values by color (row) name.
When I say that this function here can probably go away, I'm thinking of just replacing calls to it by `gamedata->GetColor(GUIColorNames[color])` (color being one of the GUIColors values).

Reply:
Sorry I don't understand this. Now I have:

- A **std::vector<std::string>** that has all the string names on the **colors.2da table.**
- A **Enum** that has the row where the color is set.
- Using the **Enum** I use that position on the vector to retrieve the color name to use on **gamedata->GetColor**

And the proposed change will work something like this:
If I use a **std::map<GUIColors, std::string>** I will have to also have a way to convert the **Enum** to a position on the string, so I have to keep the **std::vector<std::string>** and also use the map, and then I will not need  **gamedata->GetColor** because I have the color there, but if not in the map I will call **gamedata->GetColor** ?

This is what I'm thinking, I'm not fully comprehend where the change is going or how it will work.

Reply:
> I will have to also have a way to convert the Enum to a position on the string
What string? The enum still holds the row numbers and the std::map will map these enum values to the row names. Exactly like the vector, except that the indices are properly typed (enum vs general int), so there's no need to convert the index. Or care about missing indexes (keys).